instrumentation/runtime/runtime.go

```diff
@@ -97,6 +50,167 @@ func Start(opts ...Option) error {
+		"go.config.gogc",
+		metric.WithUnit("%"),
+		metric.WithDescription("Heap size target percentage configured by the user, otherwise 100."),
+	)
+	if err != nil {
+		return err
+	}
+
+	otherMemoryOpt := metric.WithAttributeSet(
+		attribute.NewSet(attribute.String("go.memory.type", "other")),
+	)
+	stackMemoryOpt := metric.WithAttributeSet(
+		attribute.NewSet(attribute.String("go.memory.type", "stack")),
+	)
+	collector := newCollector(c.MinimumReadMemStatsInterval)
+	var lock sync.Mutex
+	_, err = meter.RegisterCallback(
+		func(ctx context.Context, o metric.Observer) error {
+			lock.Lock()
+			defer lock.Unlock()
+			collector.refresh()
+			stackMemory := collector.get(goHeapMemory)
+			o.ObserveInt64(memoryUsedInstrument, stackMemory, stackMemoryOpt)
+			totalMemory := collector.get(goTotalMemory) - collector.get(goMemoryReleased)
+			otherMemory := totalMemory - stackMemory
+			o.ObserveInt64(memoryUsedInstrument, otherMemory, otherMemoryOpt)
+			// Only observe the limit metric if a limit exists
+			if limit := collector.get(goMemoryLimit); limit != math.MaxInt64 {
+				o.ObserveInt64(memoryLimitInstrument, limit)
+			}
+			o.ObserveInt64(memoryAllocatedInstrument, collector.get(goMemoryAllocated))
+			o.ObserveInt64(memoryAllocationsInstrument, collector.get(goMemoryAllocations))
+			o.ObserveInt64(memoryGCGoalInstrument, collector.get(goMemoryGoal))
+			o.ObserveInt64(goroutineCountInstrument, collector.get(goGoroutines))
+			o.ObserveInt64(processorLimitInstrument, collector.get(goMaxProcs))
+			o.ObserveInt64(gogcConfigInstrument, collector.get(goConfigGC))
+			return nil
+		},
+		memoryUsedInstrument,
+		memoryLimitInstrument,
+		memoryAllocatedInstrument,
+		memoryAllocationsInstrument,
+		memoryGCGoalInstrument,
+		goroutineCountInstrument,
+		processorLimitInstrument,
+		gogcConfigInstrument,
+	)
+	if err != nil {
+		return err
+	}
+	// TODO (#5655) support go.schedule.duration
 	return nil
 }
+
+// These are the metrics we actually fetch from the go runtime.
+var runtimeMetrics = []string{
+	goTotalMemory,
+	goMemoryReleased,
+	goHeapMemory,
+	goMemoryLimit,
+	goMemoryAllocated,
+	goMemoryAllocations,
+	goMemoryGoal,
+	goGoroutines,
+	goMaxProcs,
+	goConfigGC,
+}
+
+type goCollector struct {
+	lastCollect     time.Time
+	minimumInterval time.Duration
+	// sampleBuffer is populated by runtime/metrics
+	sampleBuffer []metrics.Sample
+	// sampleMap allows us to easily get the value of a single metric
+	sampleMap map[string]*metrics.Sample
+}
+
+func newCollector(minimumInterval time.Duration) *goCollector {
+	g := &goCollector{
+		sampleBuffer:    make([]metrics.Sample, 0, len(runtimeMetrics)),
+		sampleMap:       make(map[string]*metrics.Sample, len(runtimeMetrics)),
+		minimumInterval: minimumInterval,
+	}
+	for _, runtimeMetric := range runtimeMetrics {
+		g.sampleBuffer = append(g.sampleBuffer, metrics.Sample{Name: runtimeMetric})
+		g.sampleMap[runtimeMetric] = &g.sampleBuffer[len(g.sampleBuffer)-1]
```

Comment:
```suggestion
		// We are referencing the position of the slice. 
		// Any growth to this `sampleBuffer` slice would break the implementation.
		g.sampleBuffer = append(g.sampleBuffer, metrics.Sample{Name: runtimeMetric})
		g.sampleMap[runtimeMetric] = &g.sampleBuffer[len(g.sampleBuffer)-1]
```

Should we add a comment here to declare that we are referencing the position of the slice and that any growth to this `sampleBuffer` slice would break the implementation?

Reply:
added a similar disclaimer.